### PR TITLE
Back-compat: fix self-suicide: a:fail => !a

### LINKED ISSUE #4968: Suicide trigger not handled correctly
This `suite.rc` file runs fine with Cylc7:
```
[scheduling]
    [[dependencies]]
        graph = """
            foo => bar
            foo:fail => !foo & !bar
            foo:fail | bar => baz
            """
[runtime]
    [[foo, bar]]
        script = false
    [[baz]]
        script = sleep 10; true
```
However, with Cylc 8 task `foo` somehow gets run twice.
Relevant log entries:
```
2022-07-06T17:22:29+01:00 CRITICAL - [1/foo running job:01 flows:1] (received)failed/ERR at 2022-07-06T17:22:29+01:00
2022-07-06T17:22:29+01:00 INFO - [1/foo running job:01 flows:1] => failed
2022-07-06T17:22:29+01:00 CRITICAL - [1/foo failed job:01 flows:1] failed
2022-07-06T17:22:29+01:00 INFO - [1/foo failed job:01 flows:1] => waiting
2022-07-06T17:22:29+01:00 INFO - [1/foo waiting job:01 flows:1] => waiting(queued)
2022-07-06T17:22:29+01:00 CRITICAL - [1/foo waiting(queued) job:01 flows:1] task proxy removed (suicide)
2022-07-06T17:22:29+01:00 WARNING - [1/foo waiting(queued) job:01 flows:1] did not complete required outputs: ['submitted', 'succeeded']
2022-07-06T17:22:30+01:00 INFO - [1/foo waiting(queued) job:01 flows:1] => waiting
2022-07-06T17:22:30+01:00 INFO - Queue released: 1/foo
2022-07-06T17:22:30+01:00 WARNING - Task 1/foo already spawned in {1}
2022-07-06T17:22:30+01:00 INFO - [1/foo waiting job:01 flows:1] => preparing
2022-07-06T17:22:31+01:00 INFO - [1/foo preparing job:02 flows:1] (internal)submitted at 2022-07-06T17:22:31+01:00
2022-07-06T17:22:31+01:00 INFO - [1/foo preparing job:02 flows:1] submitted to localhost:background[51558]
2022-07-06T17:22:31+01:00 INFO - [1/foo preparing job:02 flows:1] => submitted
```
For some reason, `foo` goes back to waiting after failing.
It then gets suicided but still gets submitted a second time.
Once submitted, no messages are logged from job 2 (and there's nothing about any discarded messages).

### PR DESCRIPTION
**Pull request by hjoliver**
These changes close #4968

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` and `conda-environment.yml`.
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [x] No documentation update required.

### CODE REVIEW

**Comment by hjoliver** [comment]
OK the fixes didn't break any existing tests :sweat_smile: ... now to add a couple of new tests.

**Comment by wxtim** [comment]
On a purely pragmatic test - this appears to fix the workflow of the person who brought the issue to our attention.

File: cylc/flow/task_pool.py [hunk 1]
```diff
@@ -1169,7 +1169,8 @@ def spawn_on_output(self, itask, output, forced=False):
                 self._get_hidden_task_by_id(c_taskid)
                 or self._get_main_task_by_id(c_taskid)
             )
-            if c_task is not None:
+            if c_task is not None and c_task != itask:
+                # (Avoid self-suicide: a:fail => !a)
                 # Child already exists, update it.
```
Comment (wxtim) [inline cylc/flow/task_pool.py:1174]:
```suggestion
                # Child already exists, update it.
                # (unless trying to remove itself: a:fail => !a)
```

File: cylc/flow/task_pool.py [hunk 2]
```diff
@@ -1181,7 +1182,11 @@ def spawn_on_output(self, itask, output, forced=False):
                 )
                 # self.workflow_db_mgr.process_queued_ops()
 
-            elif (itask.flow_nums or forced) and not itask.flow_wait:
+            elif (
```
Comment (wxtim) [inline cylc/flow/task_pool.py:1185]: IMO these conditions are complex enough that a comment would be nice.

**Review by wxtim — APPROVED** [review]
- [x] Read Change (Made a couple of suggestions about documentation.
- [x] Tests seem to cover the matter, and pass!
- [x] Works on my copy of the customer's workflow.

File: tests/functional/spawn-on-demand/16-c7backcompat-self-suicide/suite.rc [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# suite.rc: Cylc 8 back-compat mode.
+
+# GitHub cylc-flow #4968: self-induced suicide should not retrigger foo below.
+[scheduler]
+    [[events]]
+        stall timeout = PT10S
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply tests/functional/spawn-on-demand/16-c7backcompat-self-suicide/suite.rc:6]: Oops, no, should be PT0S

File: cylc/flow/task_pool.py (hunk 2 above)
Comment (hjoliver) [reply cylc/flow/task_pool.py:1185]: will do ...

Comment (hjoliver) [reply cylc/flow/task_pool.py:1185]: Done.

**Review by MetRonnie — APPROVED** [review]

**Pull request merged**